### PR TITLE
Run queries for sql data sources for panels with multiple queries concurrently

### LINKED ISSUE #11711: [Feature Request] allow configuration of MaxOpenConns/MaxIdleConns/SetConnMaxLifetime for datasources
It would be great if the amount of connections to a database (mysql) connection store could be configured.
Currently ( 3a48ea8 ) both MaxIdleConns and MaxOpenConns are hardcoded as 10.
In our case this exhausts the maximum amount of allowed connections on the DB server, starving another application.

The same configuration is possible for the sqlstore, backing users, dashboards etc.
See #7427

Please include this information:
Version: 5.0.4, diagnosed with source of  3a48ea8.
Datasource: mysql
OS: Debian 8.10

### What did you do?

Configured a Mysql datasource, used it in a dashboard, opened the dashboard multiple times.
All connections to the DB were exhausted.

### What you hoped for

A setting to configure the amount of used connections

### PR DESCRIPTION
**Pull request by connection-reset**
Run queries for sql data sources for panels with multiple queries concurrently.
Also make max connections configurable in the data source settings as a way to limit the number of concurrent queries.

Fixes #11711

### CODE REVIEW

File: pkg/tsdb/sql_engine.go [hunk 1]
```diff
@@ -98,8 +98,11 @@ var NewSqlQueryEndpoint = func(config *SqlQueryEndpointConfiguration, rowTransfo
 		return nil, err
 	}
 
-	engine.SetMaxOpenConns(10)
-	engine.SetMaxIdleConns(10)
+	maxConns := config.Datasource.JsonData.Get("maxConnections").MustInt(0)
+	if maxConns > 0 {
+		engine.SetMaxOpenConns(maxConns)
+		engine.SetMaxIdleConns(maxConns)
+	}
```
Comment (torkelo) [inline pkg/tsdb/sql_engine.go]: this appears to change the logic here quite a bit, removing the SetMaxOpenConns and SetMaxIdelConns calls

**Review by torkelo — COMMENTED** [review]
@svenklemm is this something you can review?

File: pkg/tsdb/sql_engine.go (hunk 1 above)
Comment (marefr) [reply pkg/tsdb/sql_engine.go]: Agree, it was 10 earlier for both SetMaxOpenConns and SetMaxIdleConns.

**Comment by marefr** [comment]
@connection-reset thank you for the contribution. For us to accept this PR we would like configuration functionality of max open connections, max idle connections and connection max life time requested in #11711. There must also be support for configuring max open connections = 0, i.e. not set, and possibility to configure max open connections and max idle connections independently.

There must also be some sane defaults and I suggest a combination of existing (for backward compatibility) and Grafana config:
* max open connections = 10
* max idle connections = 10 
* connection max life time = 14400 (seconds)

Possibly we can change all the defaults to Grafana's default config - what do you think?

Any questions/thoughts?

**Comment by connection-reset** [comment]
What's the significance of 10 open/idle connections?
I'd rather keep the defaults from the sql package, unlimited open conns, 2 idle conns, unlimited lifetime.

**Comment by marefr** [comment]
Good question. It was selected as the default historically. Fine to use defaults from the sql package and we can add a disclaimer about it in the changelog later :+1:

File: public/app/plugins/datasource/mssql/partials/config.html [hunk 2]
```diff
@@ -29,6 +29,24 @@ <h3 class="page-heading">MSSQL Connection</h3>
 	</div>
 </div>
 
+<div class="gf-form-group">
+  <div class="gf-form max-width-30">
+    <span class="gf-form-label width-11">Max open connections</span>
+    <input type="number" min="1" class="gf-form-input" ng-model="ctrl.current.jsonData.maxOpenConns" placeholder="unlimited"></input>
```
Comment (marefr) [inline public/app/plugins/datasource/mssql/partials/config.html]: `min="1"` for all three configuration options wont allow to be set to zero once saved.

File: public/app/plugins/datasource/mssql/partials/config.html [hunk 3]
```diff
@@ -29,6 +29,24 @@ <h3 class="page-heading">MSSQL Connection</h3>
...
+  </div>
+  <div class="gf-form max-width-30">
+    <span class="gf-form-label width-11">Max idle connections</span>
+    <input type="number" min="1" class="gf-form-input" ng-model="ctrl.current.jsonData.maxIdleConns" placeholder="2"></input>
+  </div>
+  <div class="gf-form max-width-30">
+    <span class="gf-form-label width-11">Connection lifetime</span>
+    <input type="number" min="1" class="gf-form-input" ng-model="ctrl.current.jsonData.connMaxLifetime" placeholder="unlimited"></input>
```
Comment (marefr) [inline public/app/plugins/datasource/mssql/partials/config.html]: Default should be 14400 seconds and not unlimited.

**Comment by connection-reset** [comment]
I just noticed I'm handling max idle conns wrong

File: pkg/tsdb/sql_engine.go [hunk 4]
```diff
@@ -98,8 +98,18 @@ var NewSqlQueryEndpoint = func(config *SqlQueryEndpointConfiguration, rowTransfo
 		return nil, err
 	}
 
-	engine.SetMaxOpenConns(10)
-	engine.SetMaxIdleConns(10)
+	maxOpenConns := config.Datasource.JsonData.Get("maxOpenConns").MustInt(0)
+	if maxOpenConns > 0 {
+		engine.SetMaxOpenConns(maxOpenConns)
+	}
+	maxIdleConns := config.Datasource.JsonData.Get("maxIdleConns").MustInt(0)
```
Comment (marefr) [inline pkg/tsdb/sql_engine.go]: Should be `MustInt(2)` to handle defaults conrrectly.

File: pkg/tsdb/sql_engine.go [hunk 5]
```diff
@@ -98,8 +98,18 @@ var NewSqlQueryEndpoint = func(config *SqlQueryEndpointConfiguration, rowTransfo
...
+	if maxIdleConns > 0 {
+		engine.SetMaxIdleConns(maxIdleConns)
+	}
+	connMaxLifetime := config.Datasource.JsonData.Get("connMaxLifetime").MustInt(0)
```
Comment (marefr) [inline pkg/tsdb/sql_engine.go]: Should be `MustInt(14400)` to handle defaults conrrectly.

**Review by marefr — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/mssql/partials/config.html (hunk 2 above)
Comment (connection-reset) [reply public/app/plugins/datasource/mssql/partials/config.html]: 0 is the same as the sql package's default, ie leaving it empty

**Comment by connection-reset** [comment]
Regarding your change requests: Is it fine to use the sql packages defaults or not?

**Comment by marefr** [comment]
Sorry I was referring to Grafana's [default connection settings](https://github.com/grafana/grafana/blob/master/conf/defaults.ini#L83). It's easier from a Grafana maintainability standpoint to always set the defaults in case third party libraries are making changes.

**Comment by connection-reset** [comment]
Ok so the defaults are now unlimited open conns, 2 idle conns, 1440 seconds lifetime.
I squashed all the configuration related commits.

**Comment by connection-reset** [comment]
I forgot a 0

**Comment by marefr** [comment]
Great. Except you set 1440 seconds lifetime per default, not 14400 seconds.

**Comment by connection-reset** [comment]
Sorry, I forgot to commit

**Comment by marefr** [comment]
Made some ux changes to datasource configuration plus updated help tooltips to include more information, Also updated documentation.

**Review by marefr — APPROVED** [review]

**Comment by marefr** [comment]
Will wait with merging this until v5.3 stable is out so that we don't have to add version disclosure information in multiple places in the v5.3 documentation.

**Comment by marefr** [comment]
With the changes in this PR, this is how the configuration page looks:

image

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Pull request merged**